// discord.js/src/index.js
'use strict';

const Util = require('discord.js/src/util/Util');

module.exports = {
  // "Root" classes (starting points)
  BaseClient: require('discord.js/src/client/BaseClient'),
  Client: require('discord.js/src/client/Client'),
  Shard: require('discord.js/src/sharding/Shard'),
  ShardClientUtil: require('discord.js/src/sharding/ShardClientUtil'),
  ShardingManager: require('discord.js/src/sharding/ShardingManager'),
  WebhookClient: require('discord.js/src/client/WebhookClient'),

  // Utilities
  ActivityFlags: require('discord.js/src/util/ActivityFlags'),
  BitField: require('discord.js/src/util/BitField'),
  Collection: require('discord.js/src/util/Collection'),
  Constants: require('discord.js/src/util/Constants'),
  DataResolver: require('discord.js/src/util/DataResolver'),
  BaseManager: require('discord.js/src/managers/BaseManager'),
  DiscordAPIError: require('discord.js/src/rest/DiscordAPIError'),
  HTTPError: require('discord.js/src/rest/HTTPError'),
  MessageFlags: require('discord.js/src/util/MessageFlags'),
  Intents: require('discord.js/src/util/Intents'),
  Permissions: require('discord.js/src/util/Permissions'),
  Speaking: require('discord.js/src/util/Speaking'),
  Snowflake: require('discord.js/src/util/Snowflake'),
  SnowflakeUtil: require('discord.js/src/util/Snowflake'),
  Structures: require('discord.js/src/util/Structures'),
  SystemChannelFlags: require('discord.js/src/util/SystemChannelFlags'),
  UserFlags: require('discord.js/src/util/UserFlags'),
  Util: Util,
  version: require('discord.js/package.json').version,

  // Managers
  ChannelManager: require('discord.js/src/managers/ChannelManager'),
  GuildChannelManager: require('discord.js/src/managers/GuildChannelManager'),
  GuildEmojiManager: require('discord.js/src/managers/GuildEmojiManager'),
  GuildEmojiRoleManager: require('discord.js/src/managers/GuildEmojiRoleManager'),
  GuildMemberManager: require('discord.js/src/managers/GuildMemberManager'),
  GuildMemberRoleManager: require('discord.js/src/managers/GuildMemberRoleManager'),
  GuildManager: require('discord.js/src/managers/GuildManager'),
  ReactionManager: require('discord.js/src/managers/ReactionManager'),
  ReactionUserManager: require('discord.js/src/managers/ReactionUserManager'),
  MessageManager: require('discord.js/src/managers/MessageManager'),
  PresenceManager: require('discord.js/src/managers/PresenceManager'),
  RoleManager: require('discord.js/src/managers/RoleManager'),
  UserManager: require('discord.js/src/managers/UserManager'),

  // Shortcuts to Util methods
  discordSort: Util.discordSort,
  escapeMarkdown: Util.escapeMarkdown,
  fetchRecommendedShards: Util.fetchRecommendedShards,
  resolveColor: Util.resolveColor,
  resolveString: Util.resolveString,
  splitMessage: Util.splitMessage,

  // Structures
  Application: require('discord.js/src/structures/interfaces/Application'),
  Base: require('discord.js/src/structures/Base'),
  Activity: require('discord.js/src/structures/Presence').Activity,
  APIMessage: require('discord.js/src/structures/APIMessage'),
  BaseGuildEmoji: require('discord.js/src/structures/BaseGuildEmoji'),
  CategoryChannel: require('discord.js/src/structures/CategoryChannel'),
  Channel: require('discord.js/src/structures/Channel'),
  ClientApplication: require('discord.js/src/structures/ClientApplication'),
  get ClientUser() {
    // This is a getter so that it properly extends any custom User class
    return require('discord.js/src/structures/ClientUser');
  },
  Collector: require('discord.js/src/structures/interfaces/Collector'),
  DMChannel: require('discord.js/src/structures/DMChannel'),
  Emoji: require('discord.js/src/structures/Emoji'),
  Guild: require('discord.js/src/structures/Guild'),
  GuildAuditLogs: require('discord.js/src/structures/GuildAuditLogs'),
  GuildChannel: require('discord.js/src/structures/GuildChannel'),
  GuildEmoji: require('discord.js/src/structures/GuildEmoji'),
  GuildMember: require('discord.js/src/structures/GuildMember'),
  GuildPreview: require('discord.js/src/structures/GuildPreview'),
  GuildTemplate: require('discord.js/src/structures/GuildTemplate'),
  Integration: require('discord.js/src/structures/Integration'),
  Invite: require('discord.js/src/structures/Invite'),
  Message: require('discord.js/src/structures/Message'),
  MessageAttachment: require('discord.js/src/structures/MessageAttachment'),
  MessageCollector: require('discord.js/src/structures/MessageCollector'),
  MessageEmbed: require('discord.js/src/structures/MessageEmbed'),
  MessageMentions: require('discord.js/src/structures/MessageMentions'),
  MessageReaction: require('discord.js/src/structures/MessageReaction'),
  NewsChannel: require('discord.js/src/structures/NewsChannel'),
  PermissionOverwrites: require('discord.js/src/structures/PermissionOverwrites'),
  Presence: require('discord.js/src/structures/Presence').Presence,
  ClientPresence: require('discord.js/src/structures/ClientPresence'),
  ReactionCollector: require('discord.js/src/structures/ReactionCollector'),
  ReactionEmoji: require('discord.js/src/structures/ReactionEmoji'),
  RichPresenceAssets: require('discord.js/src/structures/Presence').RichPresenceAssets,
  Role: require('discord.js/src/structures/Role'),
  StoreChannel: require('discord.js/src/structures/StoreChannel'),
  Team: require('discord.js/src/structures/Team'),
  TeamMember: require('discord.js/src/structures/TeamMember'),
  TextChannel: require('discord.js/src/structures/TextChannel'),
  User: require('discord.js/src/structures/User'),
  VoiceChannel: require('discord.js/src/structures/VoiceChannel'),
  VoiceRegion: require('discord.js/src/structures/VoiceRegion'),
  VoiceState: require('discord.js/src/structures/VoiceState'),
  Webhook: require('discord.js/src/structures/Webhook'),

  WebSocket: require('discord.js/src/WebSocket'),
};
